arroyo/processing/strategies/streaming/transform.py

```diff
@@ -225,12 +231,40 @@ def parallel_transform_worker_initializer(
         custom_initialize_func()
 
 
+@dataclass
+class ParallelTransformResult(Generic[TTransformed]):
+    """
+    Results of applying the transform function to a batch
+    of messages in parallel.
+
+    `index_processed_til` holds the index of the next
+    message to be processed from the input batch. If
+    something goes wrong while processing the batch and
+    that message needs to be reprocessed, the index can
+    be passed back to the function to start processing
+    from there.
+
+    `output_batch` contains all successfully processed
+    messages.
+
+    `bad_messages` contains `InvalidMessage` exceptions
+    caught during processing. These messages (exceptions)
+    should be passed to a DLQ by raising the
+    `InvalidBatchedMessages` exception.
+    """
+
+    index_processed_til: int
+    output_batch: MessageBatch[TTransformed]
+    bad_messages: MutableSequence[InvalidMessage]
+
+
 def parallel_transform_worker_apply(
     function: Callable[[Message[TPayload]], TTransformed],
     input_batch: MessageBatch[TPayload],
     output_block: SharedMemory,
     start_index: int = 0,
-) -> Tuple[int, MessageBatch[TTransformed]]:
+    bad_messages: MutableSequence[InvalidMessage] = [],
```

Comment:
Please do not pass a mutable object to a function to simulate a different return type.
Besides making the code extremely hard to understand, this list will never be reset if you use the default parameters.
If you need to return a different type than what you returned before, change the return type.

Reply:
I'm not sure what you mean, afaik `MutableSequence` is just a regular `Sequence` but refers to the list itself being mutable so `mypy` doesn't complain when I use the `.append()` method of the class. The input type and return type are still the same, it's a list of `InvalidMessage`.

The point of this was to be able to pass the partial list of bad messages back into this function in case the batch wasn't fully processed, just like we pass a `start_index` back in when rerunning on the same batch part way through.

Reply:
Why do you need to mutate the array that is passed in though? And why does it need to be passed at all - shouldn't it always be empty initially anyway?

Reply:
Same reason `start_index` exists, if the input batch is partially processed, then next iteration an index is passed back in to know where to start processing again from.

So if the input batch is partially processed, then the array of bad messages is potentially not complete yet either and must be passed back in to continue being filled in case of bad messages in the rest of the batch.

Reply:
Then it seems we are having the wrong interface here. 
This method is called with an input batch, an index and is supposed to return a batch of processed message and a list of errors. 
For consistency, if you provide the list of invalid messages to append new errors to, you should also provide a batch of successful message to fill in so that you have a full batch of successful message and a full list of bad messages the last time you call the method.

This would be very cumbersome and counter intuitive. 
I think you should return a `ParallelTransformResult` with the batch of successful messages (as today) and the list of bad messages from the part of the batch that you processed. So the output batch is coherent with the list of bad messages. 

Then you merge the list of bad messages in `_check_for_results`

Reply:
Refactored to bring the `invalid_messages` out into the results tuple instead of passing it back into the worker.